template/base/miner.py

```diff
@@ -0,0 +1,259 @@
+import copy
+import time
+import typing
+import asyncio
+import threading
+import traceback
+
+import bittensor as bt
+
+from abc import ABC, abstractmethod
+
+# Sync calls set weights and also resyncs the metagraph.
+import template
+from template.utils.sync import sync
+from template.utils.config import check_config, add_args, config
+from template.utils.misc import ttl_get_block
+
+
+class BaseMinerNeuron(ABC):
+    """
+    Base class for Bittensor miners.
+    """
+
+    @classmethod
+    def check_config(cls, config: "bt.Config"):
+        check_config(cls, config)
+
+    @classmethod
+    def add_args(cls, parser):
+        add_args(cls, parser)
+
+    @classmethod
+    def config(cls):
+        return config(cls)
+
+    subtensor: "bt.subtensor"
+    wallet: "bt.wallet"
+    metagraph: "bt.metagraph"
+
+    def __init__(self, config=None):
+        base_config = copy.deepcopy(config or BaseMinerNeuron.config())
+        self.config = self.config()
+        self.config.merge(base_config)
+        self.check_config(self.config)
+
+        # Activating Bittensor's logging with the set configurations.
+        bt.logging(config=self.config, logging_dir=self.config.full_path)
+
+        # Warn if allowing incoming requests from anyone.
+        if not self.config.blacklist.force_validator_permit:
```

Comment:
Would we expect this config to be in every subnet miner? I can see in some situations where validator permits are not used in a particular subnet

Reply:
We think that the general case is actually that miners are not visible to outsideers - only to validators as per the docs and the way we describe these things. A subnet own can always implement specific logic such as removing this.